simple_network_sim/inference.py

```diff
@@ -0,0 +1,702 @@
+        """ Sample from prior distribution. For infection probability, we use
+        the value in the data pipeline as mean of our prior, and shape parameter
+        is fixed at 2. This defines a Beta distribution centered around our prior.
+
+        :param fitter: ABC-SMC fitter object
+        :return: New InferredInitialInfections randomly sampled from prior
+        """
+        shape = fitter.infection_probability_shape
+        kernel_sigma = fitter.infection_probability_kernel_sigma
+        mean = fitter.infection_probability.at[0, "Value"]
+        value = fitter.infection_probability.copy().assign(
+            Value=lambda x: stats.beta.rvs(shape, shape * (1 - x.Value) / x.Value, random_state=fitter.random_state))
+        return InferredInfectionProbability(value, mean, shape, kernel_sigma, fitter.random_state)
+
+    def generate_perturbated(self) -> InferredInfectionProbability:
+        """ From current parameter, add a perturbation to infection probability and return
+        a newly created perturbated parameter:
+
+        .. math::
+            P_t^* ~ K(P_t | P_{t-1}) ~ P_{t-1} + Uniform([-1,1]) * kernel_sigma
+
+        :return: New parameter which is similar to self up to a perturbation
+        """
+        value = self.value.copy()
+        value.Value = value.Value.apply(lambda x: x + self.kernel_sigma * stats.uniform
+                                        .rvs(-1., 2., random_state=self.random_state))
+        return InferredInfectionProbability(value, self.mean, self.shape, self.kernel_sigma, self.random_state)
+
+    def validate(self) -> bool:
+        """ Checks that the particle is valid, i.e. that infection probability is
+        between 0 and 1.
+
+        :return: Whether the parameter is valid
+        """
+        return np.all(self.value.Value > 0.) and np.all(self.value.Value < 1.)
+
+    def prior_pdf(self) -> float:
+        """ Compute pdf of the prior distribution evaluated at the parameter x.
+        Infection probability has a prior a beta distribution. The pdf is evaluated
+        at the current value of the parameter.
+
+        :return: pdf value of prior distribution evaluated at x
+        """
+        return stats.beta.pdf(self.value.at[0, "Value"], self.shape, self.shape * (1 - self.mean) / self.mean)
+
+    def perturbation_pdf(self, x: pd.DataFrame) -> float:
+        """ Compute pdf of the perturbation evaluated at the parameter x,
+        from the current parameter. In ABC-SMC when a particle is sampled
+        from the previous population it is slightly perturbed:
+
+        .. math::
+            P_t^* ~ K(P_t | P_{t-1}) ~ P_{t-1} + Uniform([-1,1]) * kernel_sigma
+
+        :param x: Particle to evaluate the pdf at
+        :return: pdf value of perturbation from previous particle evaluated at x
+        """
+        return stats.uniform.pdf(x.at[0, "Value"], self.value.at[0, "Value"] - self.kernel_sigma, 2 * self.kernel_sigma)
```

Comment:
Wouldn't it make sense that the column here matches the one from the InferredInitialInfections class? I think the dataframe, if pandas supported type hints, would behave sort of like a named tuple, with the column names as fields of a certain type.

Reply:
Hm not sure I understand what you mean? In both the Inferredxxxx classes I am using the dataframes as they are shaped in the data pipeline and applying transforms to them while preserving the shape.